### ISSUE TITLE
[Feature request] Add `on-find` command (mimic `ranger`'s `find` command)

### ISSUE THREAD

**Issue by Andrew15-5** [gokcehan/lf#1135]
I want to mimic `ranger`'s find command, which will auto open a directory if the `find` command successfully found one. I'm not 100% sure, but `ranger` have `set incsearch` by default (at least this feature exists). In short, I want to do this:

```lf
cmd on-find $lf -remote "send $id open"
# or
cmd on-find open # I don't know if this is a valid syntax
```

The reason is simple: this way I can press 1 less key after each `find`. Because I do this pretty often, and it became annoying to press `l` after each search pretty quickly (and there is no workaround except maybe changing the code).

Right now I can't find any way to do this. If I send `find` and `open` commands, they will be executed "simultaneously" and not "one finished (successfully), next executed".

Note: In incremental search (`find`) the selected entry may be changed, and it mustn't trigger `on-find` command. OR there could be 2 commands: `on-find` (on each entry change) `on-find-success`/`on-successful-find` and I'm interested only in the last one.

As an alternative, I propose an option that controls whether to auto-open (dir and/or file) after a successful search or not.

**Comment by Andrew15-5** [comment]
Not always, you would know at which keystroke the find command will be completed, therefore an auto-open feature will visually notify you that the find command has been completed. I tend to overshoot by keep on typing the characters which may end up being a `map` command instead of string provided to `find` command. This simple mistake can cause a huge problem if the keys you press are mapped to a destructive command.

Maybe adding an optional delay to input blocking/ignore (which is customizable in milliseconds) after `find` command was finished can be a nice little but very helpful feature.

**Comment by Andrew15-5** [comment]
I think this is all the settings I have regarding the `find` command:

```lf
set incsearch
set anchorfind!
set findlen 0
```

**Comment by Andrew15-5** [comment]
No. The only thing I found is yazi, but the shortcuts are completely different, and I don't have time to switch. I don't remember how exactly search works there.

**Comment by CatsDeservePets** [comment]
Hello @Andrew15-5 and @prabhu7x!

How about this?
```sh
cmd on-select &{{
    [ "$lf_mode" = "find" ] || exit 0
    lf -remote "send $id :open \"$f\""
}}
```

**Comment by Andrew15-5** [comment]
I'm pretty sure this also opens files, which is not desirable, as I might rename or edit the file, which are completely different actions. I guess you can modify this.

Hmm, well, I think this does indeed feel like what I wanted.

```sh
cmd on-select &{{
  [ "$lf_mode" = "find" ] || exit 0
  [ -d "$f" ] && lf -remote "send $id :open \"$f\""
}}
```

However, one edge case that I just found is that if I try to find the directory that is already selected, it will not open it. Can this be fixed as well?

**Comment by CatsDeservePets** [comment]
The problem is that `on-select` does not get executed in that case  as the selection does not change.

I tried creating a custom function or key mapping that first stores the current selection, then executes find and after that compares the previous with the current selection. However, I could not get this to work as it seems that when chaining commands, the second one gets executed without waiting `find` to finish.

Perhaps @joelim-work or @DusanLesan have some ideas.

**Comment by joelim-work** [comment]
The thing about modal commands is that there is more than one command involved:

- The original command which starts the modal command (e.g. `find`)
- Commands resulting from input to the modal command (e.g. `cmd-insert`)

Chaining from the former is quite easy:

```
map f :{{
    find
    some_other_command
}}
```

But chaining from the latter is not because it is called from many keys and is triggered internally.

---

For this particular use case, you could just add another hook command somewhere around here: https://github.com/gokcehan/lf/blob/e1016bbd748e901c4eef69f5378176a4300d46a8/eval.go#L780-L784

But I don't think it's particularly flexible. Maybe it's better to have some kind of `post-cmd` hook which executes at the end of every command - this could be used to compare selections and possibly even make other hook commands like `on-select` obsolete.

**Comment by CatsDeservePets** [comment]
> But I don't think it's particularly flexible. Maybe it's better to have some kind of `post-cmd` hook which executes at the end of every command - this could be used to compare selections and possibly even make other hook commands like on-select obsolete.

If we also added `pre-cmd` which is suggested somewhere already, we could also overrwrite built-in commands. Just as a side-note.

**Comment by Andrew15-5** [comment]
A catch-all post command does sound like a useful thing, which should fix the edge case. Not sure what you would do with prepended command (other than some cache cleaning or something?), but as a parity makes sense, though might be added on demand.

**Comment by CatsDeservePets** [comment]
> A catch-all post command does sound like a useful thing, which should fix the edge case. Not sure what you would do with prepended command (other than some cache cleaning or something?), but as a parity makes sense, though might be added on demand.

Depending on the implementation one could use shell commands to filter/modify the input before passing it to `lf`. Perhaps allow for custom env var expansion and stuff like that.

I have to think more about this besides my rather specific use cases. If it would run for every command with huge case statements to act different each time, this sounds rather slow.

**Comment by joelim-work** [comment]
So the idea of `pre-cmd`/`post-cmd` hooks are interesting, but I am also slightly concerned about the performance implications if shell commands are going to be triggerred every time a command (e.g. `down`) runs. I have not tested this however.

---

An alternative idea I had was to change modal commands like `find` so that they act as a single atomic operation, instead of entering into a different mode and wait for `cmd-insert` commands. Something like the below pseudocode:

```go
switch cmd.name {
case "find":
	key := get_keypress_from_tcell(): // block until key is pressed
	find_file(key)
	update_ui()
}
```

But it would be a completely different design and thus a large breaking change, and I'm not sure if it's worth exploring.
